### DISCUSSION TITLE
How to shave off a fixed number of pixels from a PNG with a transparent background using ImageMagick

### DISCUSSION THREAD

**Discussion by Zaggo34** [ImageMagick/ImageMagick#7234] [Help] [answered]
### ImageMagick version

 7.1.1-30 

### Operating system, version and so on

Mac OS Ventura 13.5.2 

### Description

I want to use ImageMagick to go along the outside of any shape and shave off a desired number of pixels. The shave option only applies in the x and y directions. Whereas I need it to applying along the entire edge of the shape. Does anyone know if this can be done with ImageMagick? Would really appreciate some pointers. Even if it's to a specific option. I can't seem to find anything that will do it, but it's such a powerful library that I would think it should be possible.

Thanks.

Update: I've tried to use the following:

convert input.png -alpha set -virtual-pixel transparent -channel A -morphology EdgeOut Octagon:1 +repage output.png 

Where I've set Octagon:1 for 1px, but it should be able to remove more.

But it doesn't seem to remove as much as I'd like, even if I increase the number of pixels.

<img width="637" alt="Screenshot 2024-04-09 at 13 53 21" src="https://github.com/ImageMagick/ImageMagick/assets/33038461/4bf87872-2042-4635-91c8-44459072d993">

**Comment by snibgo** [comment]
I don't understand what you want to do.

An image is always rectangular. It has a width and height. If we remove pixels, then we reduce the width or height, or both. The result is still rectangular.

Perhaps you want to change the colour or transparency of some pixels. Is that correct? Which pixels do you want to change? What do you want them to become?

**Reply by Zaggo34** [reply]
Yes you are completely right. I want to make all non transparent pixels to a chosen depth from the edge of a shape transparent. By depth here I mean a chosen length (in pixels) inwards along the perpendicular bisector of each pixel on the edge the shape.

**Comment by snibgo** [comment]
Your title says "from a PNG with a transparent background". The image you post has an alpha channel, but all pixels are fully opaque.

We can process that image like this:
```
magick twoStars.png -alpha off ( +clone -threshold 80% +write -morphology EdgeOut Disk:10 -negate ) -compose CopyOpacity -composite x.png
```
The result has turned pixels that are around the stars, within a distance of 10 pixels, transparent.

"Disk" gives a smoother edge than "octagon".

<img width="637" alt="x" src="https://github.com/ImageMagick/ImageMagick/assets/33812368/61a63cbe-268b-421f-9b95-1e161643d900">

**Reply by Zaggo34** [reply]
Hi, thanks for this. I'm sorry I made a silly mistake when uploading the picture. In fact my image does have a transparent background. I've attached it below. I also intentionally omitted another detail which seems to be affected by your command so I've included it in the attached image. That is, each image has a white border around it. 

Running your command on that image made the transparent pixels black. And also shaves pixels off the inner and outer edge of the white border. Whereas I want to preserve the transparency of the transparent pixels (of the background) and also only shave the outer edge of the white border, such that the resulting star is smaller than it originally was.

I hope that makes sense. Apologies, and thanks for your answer. This is very helpful and even if you're able to provide some further guidance it would be very much appreciated.

![stars](https://github.com/ImageMagick/ImageMagick/assets/33038461/f913d2c6-8142-46e8-801e-a7c96b98df11)

**Comment by snibgo** [comment, marked as answer]
This doesn't make transparent all of the white-ish pixels.
```
magick ^
  twoStarsB.png ^
  ( +clone ^
    -alpha extract ^
    -threshold 50%% ^
    -write mpr:ALP ^
    -morphology EdgeIn Disk:2 ^
    -negate ^
    mpr:ALP ^
    -compose Darken -composite ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  out.png
```
This is Windows BAT syntax. For bash, change every line-end caret ^ to backslash \\, and escape the parentheses \( and \), and don't double the %%.

![out](https://github.com/ImageMagick/ImageMagick/assets/33812368/7b0e0882-1ae4-44b4-ae24-423acd006b9f)

**Reply by Zaggo34** [reply]
That's worked perfectly! Thank you so much for your help!
